crates/hir-ty/src/infer.rs

```diff
@@ -258,6 +258,7 @@ pub enum InferenceDiagnostic {
 pub struct TypeMismatch {
     pub expected: Ty,
     pub actual: Ty,
+    pub ignored_in_diagnostics: bool,
```

Comment:
Instead of this, let's just add a `has_errors` field to `InferenceResult` and set that to true if there are any type mismatches pre filtering, then change mir lowering to error out if that flag is true